### ISSUE TITLE
Early reference to parameters defined via object destructuring should trigger an error

### ISSUE THREAD

**Issue by Dominator008** [microsoft/TypeScript#15274]
**TypeScript Version:**  2.2.1 / nightly (2.2.0-dev.201xxxxx)

**Code**

```ts
(function({ x = false } = { x }){})()
```

**Expected behavior:**
Give an error on trying to refer to `x` before it's defined. This is a temporal dead zone violation and throws in the browsers.
**Actual behavior:**
No error
